Enforce maxerrs more robustly

som-snytt opened the pull request:
`-Xmaxerrs` is currently advisory and depends on the good will
of the reporter. The standard console reporter is well-advised,
but others are not, such as sbt, which uses an sbt setting to
control logging.

This commit adds a simple filter in front of any reporter
deemed untrustworthy, if `-Xmaxerrs` has been set to a value
lower than the default.

som-snytt committed:
Enforce maxerrs more robustly

`-Xmaxerrs` is currently advisory and depends on the good will
of the reporter. The standard console reporter is well-advised,
but others are not, such as sbt, which uses an sbt setting to
control logging.

This commit adds a simple filter in front of any reporter
deemed untrustworthy.

Comments added as reminder that `reporters.Reporter` is
obsolete and implementations will use `internal.Reporter`.

lrytz commented on src/compiler/scala/tools/nsc/reporters/Filter.scala:
```diff
@@ -0,0 +1,54 @@
+
+/** A `Reporter` that filters events before delegating.
+ */
+abstract class Filter(delegate: Reporter) extends Reporter {
```
Could you instead create a class `DelegatingReporter`, place it next to `Reporter`, and maybe put a comment in `Reporter` somewhere mentioning the existence of the delegate? This would increase the likelyhood of the delegate to stay up to date.

som-snytt commented:
`reporters.Reporter` is scheduled for removal. I added the currently useless `DelegatingReporter` at `internal.Reporter`, where it will be useful once `global` takes an `internal.Reporter`.

I also added comments on the other reporter implementations.

lrytz commented on src/reflect/scala/reflect/internal/Reporting.scala:
```diff
@@ -116,6 +116,42 @@ abstract class Reporter {
+ *
+ *  Concrete subclasses should implement just the abstract `filter` method.
+ */
+abstract class DelegatingReporter(delegate: Reporter) extends Reporter {
```
I think keeping this unused copy of `DelegatingReporter` around is more confusing than helpful - instead a comment in `reporters.DelegatingReporter` that the class should move together with `Reporter`?

som-snytt replied:
I started to transition `Reporter` but it's nontrivial. I didn't consider just making this a mix-in. I'll try that.

lrytz commented on src/compiler/scala/tools/nsc/reporters/DelegatingReporter.scala:
```diff
@@ -0,0 +1,41 @@
+abstract class DelegatingReporter(delegate: Reporter) extends Reporter {
+
+  /* True to permit the message. */
+  protected def filter(pos: Position, msg: String, severity: Severity): Boolean
```
Should we move this method to `LimitingReporter` and make `DelegatingReporter` a pure delegate?

som-snytt replied:
OK, I didn't overthink. It was Filter before. I noticed the java tool convention is `Forwarding` as in `ForwardingJavaFileManager`. Thx for the second look.

som-snytt commented:
Changed to traits `internal.{ForwardingReporter, FilteringReporter}` to avoid hierarchy lock-in.

som-snytt commented:
I wanted to use `-Xreporter` to run vi with a list of files to edit, since I don't have an IDE with red squiggles set up, but that doesn't work under sbt.

lrytz commented on src/compiler/scala/tools/nsc/Global.scala:
```diff
@@ -75,6 +75,17 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
+  def reporter_=(newReporter: Reporter): Unit =
+    currentReporter = newReporter match {
+      case _: reporters.ConsoleReporter | _: reporters.LimitingReporter => newReporter
+      case _ if settings.maxerrs.isSetByUser && settings.maxerrs.value < settings.maxerrs.default =>
```
why do we check `settings.maxerrs.value < settings.maxerrs.default` here?

som-snytt replied:
That is a back door. Instead of a sentinel value like maxerrs=-1, if they're explicitly asking for a "huge" maxerrs, then trust their reporter to do whatever filtering and limiting it wants. Only if the user asks to "restrict" the maxerrs do we take this extraordinary step of intervening.

Probably this will never happen IRL, but for example: a reporter that removes similar errors might be trusted to respect maxerrs. Put another way, this change is really because sbt reporter doesn't respect maxerrs and when I set it to 5, I want that to work and not have to look in the sbt docs for their special setting.

lrytz replied:
OK, doppelt genäht hält besser :-)

som-snytt replied:
💯

lrytz commented on src/compiler/scala/tools/nsc/reporters/AbstractReporter.scala:
```diff
@@ -9,6 +9,8 @@ package reporters
 import scala.tools.nsc.Settings
 import scala.reflect.internal.util.Position
+// TODO
+//import scala.reflect.internal.Reporter
```
I actually think those comments are more noise than help; could you remove them across the PR? When we finally get around to removing class `nsc.reporters.Reporter`, it will be simple enough to find the places where we need to import `reflect.internal.Reporter`.

lrytz commented on src/compiler/scala/tools/nsc/reporters/StoreReporter.scala:
```diff
@@ -8,22 +8,23 @@ package reporters
   }
   val infos = new mutable.LinkedHashSet[Info]
-  protected def info0(pos: Position, msg: String, severity: Severity, force: Boolean) {
+  override protected def info0(pos: Position, msg: String, severity: Severity, force: Boolean): Unit = {
```
I think we usually don't put `override` when only implementing an abstract method.

som-snytt replied:
OK. I'll also normalize NoReporter, above. :)

lrytz reviewed (changes requested):
Just some minor cleanups.

som-snytt committed:
Review: be less annoying, more conforming
